bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+}
+
+static void
+disconnect_jack (jack_client_t *jack_client)
+{
+  assert_return (jack_client != NULL);
+
+  jack_deactivate (jack_client);
+  jack_client_close (jack_client);
+}
+
+struct DeviceDetails {
+  uint ports = 0;
+  uint input_ports = 0;
+  uint output_ports = 0;
+  uint physical_ports = 0;
+  uint terminal_ports = 0;
+  bool default_device = false;
+
+  std::vector<std::string> input_port_names;
+  std::vector<std::string> output_port_names;
+};
+
+static std::map<std::string, DeviceDetails>
+query_jack_devices (jack_client_t *jack_client)
+{
+  std::map<std::string, DeviceDetails> devices;
+
+  assert_return (jack_client, devices);
+
+  const char **jack_ports = jack_get_ports (jack_client, NULL, NULL, 0);
+  if (jack_ports)
+    {
+      bool have_default_device = false;
+
+      for (uint i = 0; jack_ports[i]; i++)
+	{
+	  const char *end = strchr (jack_ports[i], ':');
+	  if (end)
+	    {
+              std::string device_name (jack_ports[i], end);
+
+	      jack_port_t *jack_port = jack_port_by_name (jack_client, jack_ports[i]);
+	      if (jack_port)
+		{
+                  const char *port_type = jack_port_type (jack_port);
+                  if (strcmp (port_type, JACK_DEFAULT_AUDIO_TYPE) == 0)
+                    {
+	              DeviceDetails &details = devices[device_name];
+	              details.ports++;
+
+                      int flags = jack_port_flags (jack_port);
+                      if (flags & JackPortIsInput)
+                        {
+                          details.input_ports++;
+                          details.input_port_names.push_back (jack_ports[i]);
+                        }
+                      if (flags & JackPortIsOutput)
+                        {
+                          details.output_ports++;
+                          details.output_port_names.push_back (jack_ports[i]);
+                        }
+                      if (flags & JackPortIsPhysical)
+                        {
+                          details.physical_ports++;
+
+                          if (!have_default_device)
+                            {
+                              /* the first device that has physical ports is the default device */
+                              details.default_device = true;
+                              have_default_device = true;
+                            }
+                        }
+                      if (flags & JackPortIsTerminal)
+                        details.terminal_ports++;
+                    }
+		}
+	    }
+	}
+      free (jack_ports);
+    }
+
+  return devices;
+}
+
+static void
+list_jack_drivers (Driver::EntryVec &entries)
+{
+  map<std::string, DeviceDetails> devices;
+  jack_client_t *jack_client = connect_jack();
+  if (jack_client)
+    {
+      devices = query_jack_devices (jack_client);
+      disconnect_jack (jack_client);
+    }
+  else
+    {
+      // should we try to generate and show an error message if connecting jack failed?
+    }
```

Comment:
No, if there's no jack, there are no devices to list. Simple. Stay silent.

Reply:
Ok, removed the else branch.